crates/polars-ops/src/series/ops/replace.rs

```diff
@@ -58,26 +64,95 @@ pub fn replace(
     }
 }
 
+/// Replace all values by different values.
+///
+/// Raises an error if not all values were replaced.
+pub fn replace_strict(
+    s: &Series,
+    old: &Series,
+    new: &Series,
+    return_dtype: Option<DataType>,
+) -> PolarsResult<Series> {
+    if old.len() == 0 {
+        polars_ensure!(
+            s.len() == s.null_count(),
+            InvalidOperation: "must specify which values to replace"
+        );
+        return Ok(s.clone());
+    }
+    validate_old(old)?;
+
+    let old = cast_old_to_series_dtype(old, s.dtype())?;
+    let new = match return_dtype {
+        Some(dtype) => new.strict_cast(&dtype)?,
+        None => new.clone(),
+    };
+
+    if new.len() == 1 {
+        replace_by_single_strict(s, &old, &new)
+    } else {
+        replace_by_multiple_strict(s, old, new)
+    }
+}
+
+/// Validate the `old` input.
+fn validate_old(old: &Series) -> PolarsResult<()> {
+    polars_ensure!(
+        old.n_unique()? == old.len(),
+        InvalidOperation: "`old` input for `replace` must not contain duplicates"
+    );
+    Ok(())
+}
+
+/// Cast `old` input while enabling String to Categorical casts.
+fn cast_old_to_series_dtype(old: &Series, dtype: &DataType) -> PolarsResult<Series> {
+    match (old.dtype(), dtype) {
+        #[cfg(feature = "dtype-categorical")]
+        (DataType::String, DataType::Categorical(_, ord)) => {
+            let empty_categorical_dtype = DataType::Categorical(None, *ord);
+            old.strict_cast(&empty_categorical_dtype)
+        },
+        _ => old.strict_cast(dtype),
+    }
+}
+
 // Fast path for replacing by a single value
 fn replace_by_single(
     s: &Series,
     old: &Series,
     new: &Series,
     default: &Series,
 ) -> PolarsResult<Series> {
-    let mask = if old.null_count() == old.len() {
-        s.is_null()
-    } else {
-        let mask = is_in(s, old)?;
-
-        if old.null_count() == 0 {
-            mask
-        } else {
-            mask.bitor(s.is_null())
-        }
-    };
+    let mut mask = get_replacement_mask(s, old)?;
+    if old.null_count() > 0 {
+        mask = mask.fill_null_with_values(true)?;
+    }
     new.zip_with(&mask, default)
 }
+/// Fast path for replacing by a single value in strict mode
+fn replace_by_single_strict(s: &Series, old: &Series, new: &Series) -> PolarsResult<Series> {
+    let mask = get_replacement_mask(s, old)?;
+    ensure_all_replaced(&mask, s, old.null_count() > 0, true)?;
+
+    let mut out = new.new_from_index(0, s.len());
+
+    // Transfer validity from `mask` to `out`.
+    if mask.null_count() > 0 {
```

Comment:
For a future PR we must optimize this. Adding a validity is almost free, but here it takes a very expensive branchy path.

Reply:
Yeah I wasn't happy about this either - but I couldn't find a way to do this without a lot of code... I figure we're probably missing some API building blocks here.